idx/query.go

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package idx
+
+import (
+	"github.com/m3db/m3ninx/search"
+	"github.com/m3db/m3ninx/search/query"
+)
+
+// Query encapsulates a search query for an index.
+type Query struct {
+	query search.Query
+}
+
+// NewTermQuery returns a new query for finding documents which match a term exactly.
+func NewTermQuery(field, term []byte) Query {
+	return Query{
+		query: query.NewTermQuery(field, term),
+	}
+}
+
+// NewRegexpQuery returns a new query finding documents which match a regular expression.
+func NewRegexpQuery(field, regexp []byte) (Query, error) {
+	q, err := query.NewRegexpQuery(field, regexp)
+	if err != nil {
+		return Query{}, err
+	}
+	return Query{
+		query: q,
+	}, nil
+}
+
+// BooleanClause represents the type of clause in a boolean query.
+type BooleanClause int
+
+// The valid types of clauses in a boolean query are Must, Should, and MustNot.
+const (
+	Must BooleanClause = iota
+	Should
+	MustNot
+)
+
+// BooleanQueryBuilder is used to build boolean queries.
+type BooleanQueryBuilder struct {
+	must    []search.Query
+	should  []search.Query
+	mustNot []search.Query
+}
+
+// Add adds a query with the given type of clause to the builder. The builder is also
+// returned so that calls to Add can be chained together.
+func (b *BooleanQueryBuilder) Add(q Query, c BooleanClause) *BooleanQueryBuilder {
```

Comment:
how about three functions instead of Add, i.e.

```
func (b *BooleanQueryBuilder) Must(q Query) *BooleanQueryBuilder {}
func (b *BooleanQueryBuilder) Should(q Query) *BooleanQueryBuilder {}
func (b *BooleanQueryBuilder) ShouldNot(q Query) *BooleanQueryBuilder {}
```